### PR TITLE
New routers

### PR DESCRIPTION
**Pull request by galhilu**
Hi David,
The changes to be reviewed are mostly in routerGenServer and include the creation of the routing table (that also uses a newly written function in NerlTools) in the init function, and the addition of 2 new cast functions: uniCast and broadCast. 
changes in other files  (mainGenServer and clientGenServer) were for testing purposes.

### CODE REVIEW

File: src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl [hunk 1]
```diff
@@ -53,7 +53,11 @@ init({MyName,NerlnetGraph}) ->
   ?LOG_NOTICE("Router ~p is connected to: ~p~n",[MyName, [digraph:vertex(NerlnetGraph,Vertex) || Vertex <- digraph:out_neighbours(NerlnetGraph,MyName)]]),
     % nerl_tools:start_connection([digraph:vertex(NerlnetGraph,Vertex) || Vertex <- digraph:out_neighbours(NerlnetGraph,MyName)]),
   put(nerlnetGraph, NerlnetGraph),
-  {ok, #router_genserver_state{msgCounter = 1, myName = MyName }}.
+  EtsRef = ets:new(rouint_table, [set]),
```
Comment (galhilu) [inline src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl:56]: routing typo

File: src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl [hunk 2]
```diff
@@ -176,9 +180,64 @@ handle_cast({getStats,_Body}, State  = #router_genserver_state{myName = MyName,
...
+handle_cast({uniCast,{Dest,Body}}, State = #router_genserver_state{msgCounter = MsgCounter,etsRef=Routing_table }) ->
+
+  [{Dest,{Name,Host,Port}}]=ets:lookup(Routing_table,Dest),
+  case Dest of
+    Name->
+      %the destination is the next hop, send as regular message
+      {Path,Data}=Body;
+    _->
+      %next hop isnt the destination, continue as geneal router message
+      Path="uniCast",
```
Comment (galhilu) [inline src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl:192]: atom to list

File: src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl [hunk 3]
```diff
@@ -176,9 +180,64 @@ handle_cast({getStats,_Body}, State  = #router_genserver_state{myName = MyName,
...
+    end,
+    maps:put({Name,Host,Port},NewVal,Acc)
+  end,
+
+  NextHopMap=lists:foldl(MapFunc,#{},DestList),
+
+  SendFunc=fun({Name,Host,Port},DestClientList)->
+    %iterate on the maps keys (addreses) and forword message according to 1 of 3 cases
+    case length(DestClientList) of
+      1->
```
Comment (galhilu) [inline src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl:218]: define

File: src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl [hunk 4]
```diff
@@ -176,9 +180,64 @@ handle_cast({getStats,_Body}, State  = #router_genserver_state{myName = MyName,
...
+        %only 1 dest for the addres, switch to unicast or lose the general format if last hop
+        [Client]=DestClientList,
```
Comment (galhilu) [inline src_erl/Communication_Layer/http_Nerlserver/src/Router/routerGenserver.erl:220]: no client in names

File: src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.erl [hunk 5]
```diff
@@ -9,6 +9,7 @@
 -export([getdeviceIP/0, port_available/1]).
 -export([list_to_numeric/1]).
 -export([calculate_size/1]).
+-export([make_rouint_table/4]).
```
Comment (leondavi) [inline src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.erl:12]: Typo

File: src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.erl [hunk 6]
```diff
@@ -188,4 +189,18 @@ calculate_size(List) when is_list(List) ->
 
 %% TODO: add another timing map for NIF of each worker action
 
-%% TODO: create create_body func for standard message passing
\ No newline at end of file
+%% TODO: create create_body func for standard message passing
+
+make_rouint_table(Ets,[],_Origin,_NerlnetGraph)->Ets;
```
Comment (leondavi) [inline src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.erl:194]: Typo

**Review by leondavi — CHANGES_REQUESTED** [review]

**Comment by leondavi** [comment]
@galhilu found typo in function name
@GuyPerets106 please review and approve

please validate that synthetic and mnist experiments pass and number of messages is correct.

@galhilu thanks for your contribution!

**Comment by leondavi** [comment]
This PR changes moved to FullFlowCI New Routers

**Pull request closed without merging**